lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -704,3 +742,201 @@ dummyProtocolParameters = ProtocolParameters
+        === estimateTxSize emptyTxSkeleton
+
+-- Tests that using 'txConstraints' to estimate the cost of a non-empty
+-- selection produces a result that is consistent with the result of using
+-- 'estimateTxCost'.
+--
+prop_txConstraints_txCost :: MockSelection -> Property
+prop_txConstraints_txCost mock =
+    counterexample ("result: " <> show result) $
+    counterexample ("lower bound: " <> show lowerBound) $
+    counterexample ("upper bound: " <> show upperBound) $
+    conjoin
+        [ result >= lowerBound
+        , result <= upperBound
+        ]
+  where
+    MockSelection {txInputCount, txOutputs, txRewardWithdrawal} = mock
+    result :: Coin
+    result = mconcat
+        [ txBaseCost mockTxConstraints
+        , txInputCount `mtimesDefault` txInputCost mockTxConstraints
+        , F.foldMap (txOutputCost mockTxConstraints . tokens) txOutputs
+        , txRewardWithdrawalCost mockTxConstraints txRewardWithdrawal
+        ]
+    lowerBound = estimateTxCost mockProtocolParameters emptyTxSkeleton
+        {txInputCount, txOutputs, txRewardWithdrawal}
+    -- We allow a small amount of overestimation due to the slight variation in
+    -- the marginal cost of an input:
+    upperBound = lowerBound <> txInputCount `mtimesDefault` Coin 8
+
+-- Tests that using 'txConstraints' to estimate the size of a non-empty
+-- selection produces a result that is consistent with the result of using
+-- 'estimateTxSize'.
+--
+prop_txConstraints_txSize :: MockSelection -> Property
+prop_txConstraints_txSize mock =
+    counterexample ("result: " <> show result) $
+    counterexample ("lower bound: " <> show lowerBound) $
+    counterexample ("upper bound: " <> show upperBound) $
+    conjoin
+        [ result >= lowerBound
+        , result <= upperBound
+        ]
+  where
+    MockSelection {txInputCount, txOutputs, txRewardWithdrawal} = mock
+    result :: TxSize
+    result = mconcat
+        [ txBaseSize mockTxConstraints
+        , txInputCount `mtimesDefault` txInputSize mockTxConstraints
+        , F.foldMap (txOutputSize mockTxConstraints . tokens) txOutputs
+        , txRewardWithdrawalSize mockTxConstraints txRewardWithdrawal
+        ]
+    lowerBound = estimateTxSize emptyTxSkeleton
+        {txInputCount, txOutputs, txRewardWithdrawal}
+    -- We allow a small amount of overestimation due to the slight variation in
+    -- the marginal size of an input:
+    upperBound = lowerBound <> txInputCount `mtimesDefault` TxSize 4
+
+newtype Large a = Large { unLarge :: a }
+    deriving (Eq, Show)
+
+instance Arbitrary (Large TokenBundle) where
+    arbitrary = fmap Large . genFixedSizeTokenBundle =<< choose (1, 128)
+
+-- Tests that if a bundle is oversized (when serialized), then a comparison
+-- between 'txOutputSize' and 'txOutputMaximumSize' should also indicate that
+-- the bundle is oversized.
+--
+prop_txConstraints_txOutputMaximumSize :: Large TokenBundle -> Property
+prop_txConstraints_txOutputMaximumSize (Large bundle) =
```

Comment:
The `bundle` argument probably be blinded to avoid spamming the logs with the output of `show` for huge token bundles. (We already have the data we need in the counterexamples.)

Reply:
Fixed in 2099f94ebad72a8bdf1bd04b7e42120694075d32.